docs/architecture/engine.md

```diff
@@ -0,0 +1,262 @@
+
+Lets imagine a read for a given series that requests the last 6 hours worth of data, and an M3DB namespace that is configured with a blocksize of 2 hours (I.E we need to find 3 different blocks.)
+
+If the current time is 8PM, then the location of the requested blocks might be as follows:
+
+(TODO: Make this a nice diagram)
+
+[2PM - 4PM (Fileset file)] - Because this block was sealed and flushed and hasn't been read recently
+
+[4PM - 6PM (In-memory cache)] - Because this block was read from disk recently and cached
+
+[6PM - 8PM (active buffer)] - Because this block hasn't been sealed and flushed to disk yet
+
+
+Then M3DB will need to consolidate:
+
+1) The not-yet-sealed block from the active buffers / encoders (located inside an internal lookup in the Series object) [6PM - 8PM]
+2) The in-memory cached block (also located inside an internal lookup in the Series object) [4PM - 6PM]
+3) The block from disk (the block retrieve from disk will then be cached according to the current [caching policy](engine.md#caching-policies) [2PM - 4PM]
+
+M3DB will retrieve the three blocks from their respective locations in memory / on-disk, and transmit all of the data back to the client. Note that since M3DB nodes return compressed blocks (the M3DB client decompresses them) its not possible to return "partial results" for a given block. If any portion of a read requests spans a given block, then that block in its entirety must be transmitted back to the client. In practice, this ends up being not much of an issue because of the high compression ratio that M3DB is able to achieve.
+
+## Caching policies
+
+Blocks that are still being actively compressed / M3TSZ encoded must be kept in memory until they are sealed and flushed to disk. Blocks that have already been sealed, however, don't need to remain in-memory. In order to support efficient reads, M3DB implements various caching policies which determine which flushed blocks are kept in memory, and which are not. The "cache" itself is not a separate datastructure in memory, cached blocks are simply stored in their respective series object.
+
+### None Cache Policy
+
+The none cache policy is the simplest. As soon as a block is sealed, its flushed to disk and never retained in memory again. This cache policy will have the lowest memory consumption, but also the poorest read performance as every read for a block that is already flushed will require a disk read.
+
+### All Cache Policy
+
+The all cache policy is the opposite of the none cache policy. All blocks are kept in memory until their retention period is over. This policy can be useful for read-heavy workloads with small datasets, but is obviously limited by the amount of memory on the host machine. Also keep in mind that this cache policy may have unintended side-effects on write throughput as keeping every block in memory creates a lot of work for the Golang garbage collector.
```

Comment:
Assuming that creating a lot of work for the Go garbage collector increases memory usage but might be good to state that explicitly